src/batch/src/executor/hash_agg.rs

```diff
@@ -227,6 +237,7 @@ impl<K: HashKey + Send + Sync> HashAggExecutor<K> {
 
                 // TODO: currently not a vectorized implementation
                 for state in states {
+                    check_shutdown(&self.shutdown_rx)?;
```

Comment:
It seems too frequent to check shutdown here. I think checking it after each chunk should be enough.

Reply:
one check per chunk +1

Reply:
e.g. a infinite loop executor. It never get the shutdown signal and it will not yield so that we can't drop this future. 

```
execute(){
    loop {
        do_some_thing()
    }
}
```

So to shutdown it, we have to check shutdown internally,
```
execute(){
    loop {
        check_shutdown()?; <----- add check here 
        do_some_thing()
    }
}
```

Reply:
> one check per chunk +1

Acutally, after #9520, we can capture the shutdown signal for each data chunk. This PR is to prevent the case that the process time of a chunk is to long. e.g.

```
for data_chunk in child_executor { <--- we can capture shutdonw signal here.
    
    // but what if the following process take a long time?
    for row in data_chunk {
        process(row)
    }

    yield data_chunk <--- we can capture shutdown signal here.
}
```

So for shutdown more timely, we add manual shutdown check internally,
```
for data_chunk in child_executor { <--- we can capture shutdonw signal here.
    
    // but what if the following process take a long time?
    for row in data_chunk {
        check_shutdown() <--------- add manual check here
        process(row)
    }

    yield data_chunk <--- we can capture shutdown signal here.
}
```
@richardchien @wangrunji0408

Reply:
> This PR is to prevent the case that the process time of a chunk is to long.

In what case will the process time be too long? IIUC chunk size is restricted to 4096 or so?

Reply:
> In what case will the process time be too long? IIUC chunk size is restricted to 4096 or so?

chunk size seems is a config var? (dafault var is 4096). 

And it based on a assume that the cost of check_shutdown is small.(CPU is not the bottleneck) So even if it's 4096, the following case can make it shutdown more timely.🤔
e.g. nest loop join
```
#[for_await]
for right_chunk in right.execute() {
    let right_chunk = right_chunk?;
    
    for left_row in left.iter().flat_map(|chunk| chunk.rows()) {
        check_shutdown(&shutdown_rx)?;  <----------------- add check here 
        
        let chunk = Self::concatenate_and_eval(
            join_expr.as_ref(),
            &left_data_types,
            left_row,
            &right_chunk,
        )
        .await?;
        
        if chunk.cardinality() > 0 {
            for spilled in chunk_builder.append_chunk(chunk) {
                yield spilled
            }
        }
    }
}
```

Reply:
> e.g. a infinite loop executor. It never get the shutdown signal and it will not yield so that we can't drop this future.
> 
> ```
> execute(){
>     loop {
>         do_some_thing()
>     }
> }
> ```
> 
> So to shutdown it, we have to check shutdown internally,
> 
> ```
> execute(){
>     loop {
>         check_shutdown()?; <----- add check here 
>         do_some_thing()
>     }
> }
> ```

Yes, but we can add [yield point](https://docs.rs/tokio/latest/tokio/task/fn.consume_budget.html) inside the loop.

```rust
async fn execute() {
    loop {
        do_some_thing();
        tokio::task::consume_budget().await; // will yield to runtime every ~100th call.
    }
}
```

Reply:
> And it based on a assume that the cost of check_shutdown is small.(CPU is not the bottleneck) So even if it's 4096, the following case can make it shutdown more timely.🤔

Yes, the overhead is not high, but it's intrusive, making code complex. What I'm concerning is that is it worthwhile to do so?


> ```rust
> async fn execute() {
>     loop {
>         do_some_thing();
>         tokio::task::consume_budget().await; // will yield to runtime every ~100th call.
>     }
> }
> ```

This LGTM. Could be less intrusive than passing `shutdown_rx` everywhere.

Reply:
Yes, I don't know this API before...🥵 It's more clear. 

`tokio::task::consume_budget()` also can guarantee the react fast enough. And we don't need to add stop flag in every executor. 

@liurenjie1024 how do you think

Reply:
> `yield_now` is much more heavy then shutdown check(an atomic load), it may cause task to enqueue/dequeue.

What's the purpose of atomic load? That is out of my expectation. 👀

Reply:
Oh I see. 😂

Reply:
So there are two manual check

`consume_budget` 

advantage:
- more clear code. We don't need to pass shutdown everywhere.

disadvantage:
- It doesn't ensure immediately yield. But it can gurantee that the shutdown signal will delay at most 100(or more less) call.

`check_shutdown`

advantage:
- immediately yield
- more lightweight

disadvantage:
- it's intrusive, making code complex

In addition, for per chunk check, I think `managed executor` is enough. We only need manual and more fine-grained check for where pre chunk process may take a long time. 

So based on the fact that there are no much long time pre-chunk process, `managed executor` + `check_shutdown` may be accpactable?  @wangrunji0408 @liurenjie1024 @richardchien